spring/boot3-autoconfigure/src/main/java/com/linecorp/armeria/internal/spring/ArmeriaConfigurationUtil.java

```diff
@@ -463,5 +465,19 @@ public static long parseDataSize(String dataSizeText) {
         }
     }
 
+    /**
+     * Remove duplicate {@link Port} entries within the list.
+     * If there are duplicates, keep the preceding {@link Port}.
+     * The criteria for duplication are based on {@link Port#port}, {@link Port#address}, and {@link Port#ip}.
+     */
+    private static List<Port> dedupPorts(List<Port> ports) {
+        return ports.stream()
+                    .collect(toImmutableMap(
+                            port -> Objects.hashCode(port.getPort(), port.getAddress(), port.getIp()),
```

Comment:
This `hashCode` could return the same value even if two ports have no same properties.
We should compare each property.
How about also just using two for loops for simplicity since the list is not very large?
```java
private static List<Port> dedupPorts(List<Port> ports) {
    final List<Port> dedupedList = new ArrayList<>();
    for (Port port : ports) {
        boolean found = false;
        for (Port deduped : dedupedList) {
            if (port.getPort() == deduped.getPort() &&
                Objects.equals(port.getAddress(), deduped.getAddress()) &&
                Objects.equals(port.getIp(), deduped.getIp())) {
                found = true;
                // We might want to merge port.getProtocols() if needed.
                break;
            }
        }
        if (!found) {
            dedupedList.add(port);
        }
    }
    return dedupedList;
}
```

Reply:
Thanks! I have pushed the code with added logic to merge `SessionProtocol` lists and convert the final result list to `ImmutableList`.

Reply:
To reduce communication costs, I tried to push the code first and seek feedback rather than asking about the direction in advance. However, I realize now that sometimes, it might have been better to ask about the direction before pushing the changes. I apologize for that. 🙇
Since I don't have much experience in contributing to open-source projects or participating in large projects, if there are any issues with my communication approach, please feel free to let me know. Thank you. 🙇

Reply:
Oh, no need to apologize for that. 😄
Coding can be tricky, and sometimes we underestimate how challenging it can be until we actually start implementing it.
I must say, you're doing an excellent job and making a significant contribution to the community. So, please don't worry about it and keep up the fantastic work! 🙇